server/src/main/java/com/microsoft/java/bs/core/internal/reporter/TestReportReporter.java

```diff
@@ -0,0 +1,193 @@
+    testDuration = 0;
+  }
+
+  private TestName getTestName(JvmTestOperationDescriptor eventDescriptor) {
+    List<JvmTestOperationDescriptor> fullStack = new ArrayList<>();
+    fullStack.add(eventDescriptor);
+    OperationDescriptor descriptor = eventDescriptor.getParent();
+    while (descriptor != null) {
+      if (descriptor instanceof JvmTestOperationDescriptor jvmTestOperationDescriptor) {
+        fullStack.add(jvmTestOperationDescriptor);
+      }
+      descriptor = descriptor.getParent();
+    }
+    // Gradle can have blank classnames on dynamic tests even though the test is still
+    // within the class, so search until classname disappears completely.
+    int i = fullStack.size() - 1;
+    boolean classNameFound = false;
+    while (i >= 0 && !classNameFound) {
+      if (fullStack.get(i).getClassName() != null) {
+        classNameFound = true;
+      } else {
+        i--;
+      }
+    }
+    // earlier check means that classname will always be found so can't have i < 0
+    // reverse list order
+    TestName testName = null;
+    while (i >= 0) {
+      JvmTestOperationDescriptor desc = fullStack.get(i);
+      TestName currentTestName = new TestName(desc.getDisplayName(), desc.getSuiteName(),
+          desc.getClassName(), desc.getMethodName());
+      currentTestName.setParent(testName);
+      testName = currentTestName;
+      i--;
+    }
+    return testName;
+  }
+
+  @Override
+  public void statusChanged(ProgressEvent event) {
+    if (client != null) {
+      if (event.getDescriptor() instanceof JvmTestOperationDescriptor descriptor) {
+        TestName testName = getTestName(descriptor);
+        // do not send reports on Gradle internal test tasks
+        if (testName != null) {
+          String taskPath = getTaskPath(event.getDescriptor());
+          TaskId taskId = getTaskId(taskPath);
+          if (event instanceof StartEvent) {
+            TaskStartParams startParam = new TaskStartParams(taskId);
+            startParam.setMessage("Start test");
+            startParam.setDataKind("test-start");
+            startParam.setEventTime(event.getEventTime());
+            TestStartEx testStart = new TestStartEx(event.getDisplayName(), testName);
+            startParam.setData(testStart);
+            client.onBuildTaskStart(startParam);
+          } else if (event instanceof FinishEvent finishEvent) {
+            OperationResult result = finishEvent.getResult();
+            testDuration += result.getEndTime() - result.getStartTime();
+            StatusCode statusCode = StatusCode.OK;
+            TestStatus testStatus = TestStatus.PASSED;
+            String stackTrace = null;
+            if (result instanceof TestFailureResult testFailureResult) {
+              statusCode = StatusCode.ERROR;
+              testStatus = TestStatus.FAILED;
+              stackTrace = testFailureResult.getFailures()
+                  .stream()
+                  .filter(f -> f instanceof DefaultTestAssertionFailure)
```

Comment:
Only extract the stackTrace from DefaultTestAssertionFailure  may not be enough. Take this **junit4** example:

```java
public class ExceptionInBefore {

    @BeforeClass  
    public static void beforeClass() throws Exception {
        throw new Exception("Exception in @BeforeClass");
    };

    @Test
    public void test() {
        
    }
}
```

In this case, the stackTrace is stored in DefaultTestFrameworkFailure

Reply:
I've pushed an update to use `TestFailure` instead of `DefaultTestAssertionFailure` and added a test using `@BeforeAll`